File: most_frequent_subtree_sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var findFrequentTreeSum = function(root) {
    const freq = {};

    function dfs(node) {
        if (!node) {
            return 0;
        }

        // calculate the subtree sum recursively
        const leftSum = dfs(node.left);
        const rightSum = dfs(node.right);

        const subtreeSum = leftSum + rightSum + node.val;
        // increment the frequency of the subtree sum
        freq[subtreeSum] = 1 + (freq[subtreeSum] || 0);

        return subtreeSum;
    }

    dfs(root);

    const maxFreq = Math.max(...Object.values(freq));
    return Object.entries(freq)
        .filter(([sum, freq]) => freq === maxFreq)
        .map(([sum, freq]) => parseInt(sum));    
};

File: most_frequent_subtree_sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> findFrequentTreeSum(TreeNode* root) {
        unordered_map<int, int> freq;
        dfs(root, freq);
        
        int maxFreq = 0;
        for (auto& [sum, count] : freq) {
            maxFreq = max(maxFreq, count);
        }
        
        vector<int> res;
        for (auto& [sum, count] : freq) {
            if (count == maxFreq) {
                res.push_back(sum);
            }
        }
        
        return res;        
    }

private:
    int dfs(TreeNode* node, unordered_map<int, int>& freq) {
        if (node == nullptr) {
            return 0;
        }
        
        // calculate the subtree sum recursively
        int leftSum = dfs(node->left, freq);
        int rightSum = dfs(node->right, freq);
        int subtreeSum = leftSum + rightSum + node->val;
        
        // increment the frequency of the subtree sum
        freq[subtreeSum]++;

        return subtreeSum;
    }    
};
